Add support to periodic allocations

MatteoDrago opened the pull request:
This is a basic proposal to implement periodic allocations into our version of the scheduler. The code is backward compatible, i.e., also non-periodic requests are considered and the remaining time gaps on the DTI are considered as CBAP time. Where possible, all for loops have been substituted by range-based for loops.

IMHO the code is well documented (and I wrote meaningful commit messages, hopefully) so it shouldn't be necessary to explain line-by-line what I coded. However, there are few assumptions that need to be considered to clarify what is supported and what's not:

- `m_availableSlots` is now the key attribute for managing all the requests.
- Multiple BI periodicity is not supported.
- For existing allocations, `ModifyExistingAllocation` considers only the duration; as of now, only decreasing the current duration is possible, while the increase is not supported by this version of the scheduler.
- `AdjustExistingAllocation` considers just the update of `m_availableSlots`, based on changes to `m_addtsAllocationList` carried out inside `DmgWifiScheduler`.
- As of now, a periodic allocation is accepted only if the scheduler can guarantee at least 2 blocks inside the DTI (in our realization of the scheduler, 1 block = 1 SP).

The code has been tested with different combinations to check if everything was allocated correctly. **However**, when and if it will be used, I suggest to enable the logs and double check if it works properly. I worked hard to try to cover all the possible conditions, but still.

MatteoDrago committed:
Init class to manage requests for periodic allocations

MatteoDrago committed:
Draft version of UpdateStartAndRemainingTime

MatteoDrago committed:
Add simple version of AddNewAllocation and introduced UpdateAvailableSlots

This version of AddNewAllocation implements a periodic-only acceptance 
policy. The assumption is that ADDTS requests include only periodic 
traffic. The member variable m_availableSlots has been introduced to 
manage this new type of traffic, as the previous variables were not 
enough. For this reason, a specific method has been created to update 
this variable.

MatteoDrago committed:
Create draft version of AddBroadcastCbapAllocations

In this version, the method fills the remaining available slots with 
broadcast CBAPs.

MatteoDrago committed:
Simplify AddBroadcastCbapAllocations

Now the Cbap Allocations are added based on m_allocationLists and 
m_remainingTime

MatteoDrago committed:
Fixed condition on UpdateAvailableSlots and refactored input parameter

MatteoDrago committed:
Draft implementation of allocation of non-periodic SPs

The requested SP is scheduled in a portion of one of the available 
slots.

MatteoDrago committed:
Fixed spStart update to follow standard guidelines

MatteoDrago committed:
Implemented VerifyAvailableSLots.

This method checks if it is possible to insert the required periodic 
allocation, based on the available time slots in the DTI.

MatteoDrago committed:
Simplified AddBroadcastCbapAllocations

In this version of the method we simply fill each available time slot in 
the DTI with a CBAP. At the end, we order the m_allocationList for 
possible further future improvements.

MatteoDrago committed:
Update doxy for UpdateAvailableSlots

MatteoDrago committed:
Update AddNewAllocation to include more than one periodic allocation.

This version of AddNewAllocation allows the scheduling of more than one periodic allocation in the DTI. Also, non-periodic allocations can be allocated either before or after the periodic ones.

MatteoDrago committed:
Optimized AddBroadcastCbapAllocation with range-based for loops

MatteoDrago committed:
Add comments to AddNewAllocation and fix indentation

MatteoDrago committed:
Modify UpdateAvailableSlots to consider changes to existing allocations

MatteoDrago committed:
Modify UpdateStartAndRemainingTime to keep it consistent

Considering how we manage the periodic requests, this method is not of key importance. However, to keep its consistency we update m_remainingDtiTime and we initialize m_availableSlots if no allocations have been requested yet.

MatteoDrago committed:
Change VerifyAvailableSlots to GetAvailableBlocks and add comments

MatteoDrago committed:
Add ModifyExistingAllocation with inline description of how it works

MatteoDrago committed:
Add AdjustExistingAllocation with proper documentation

As of now the input parameters passed to this method are useless. Being m_availableSlots of key importance to manage the allocations, this method simply update the list of available slots. This is a consequence of the fact that AdjustExistingAllocations is triggered after a DELTS request or after a clean up procedure of all the non-pseudostatic allocations.

MatteoDrago committed:
Removed m_minBroadcastCbapDuration and m_interAllocationDistance

MatteoDrago committed:
Applied ./utils/check-style.py

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+  // This method is called upon a DelTsRequest or after the cleanup of 
+  // non-pseudostatic allocations.
+  // In this version of the periodic scheduler, existing allocations are not shifted 
+  // to fill the created gaps but only the vector listing the available slots is updated.
```
"to fill the created gaps only the vector ... "

MatteoDrago replied:
I don't think it is correct, as it is a continuation of the sentence above.

tommyazz replied:
My bad :)

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+  auto addtsListCopy = m_addtsAllocationList;
+
+  // sort the copy to simplify the process of going through the allocation list
+  sort (addtsListCopy.begin (),
```
Why a sort over the `addtsAllocationList` is needed here? Should the allocation list be ordered by AllocationStart by construction?

MatteoDrago replied:
As I noticed while programming, `m_addtsAllocationList` could be unordered (it depends on the admission and scheduling policy). Since `m_availableSlot` **must be ordered** to work properly on each method that makes use of it, we need to go through `addtsAllocationList` by increasing `AllocationStart` time. 
I hope this is what you meant, otherwise please let me know!

tommyazz replied:
I can't figure out why `m_addtsAllocationList` could be unordered. Because in the end you are filling `m_addtsAllocationList` given the available slots in `m_availableSlot` which should be ordered right?
Maybe we should discuss this offline

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+    });
+
+  uint32_t startSlot = 0;
+  std::vector<std::pair<uint32_t, uint32_t> > newDTI; 
```
should be `newDti`

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+      newDTI.push_back (std::make_pair (startSlot,  m_dtiDuration.GetMicroSeconds ()));
+    }
+
+  m_availableSlots = newDTI;
```
What if you clear `m_availableSlots` before the for-cycle instead of creating and modifying a vector of the same type?

MatteoDrago replied:
That's actually a good point, basically as I did in `AdjustExistingAllocation`. Initially I did it to avoid problems in case of concurrent access to `m_availableSlots`, but since we the scheduler is a single entity, we shouldn't have the problem. 

If also @mattia-lecci agrees, I'll do it as you suggested to avoid unnecessary memory occupation :)

mattia-lecci replied:
agreed 👍

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,108 @@
+   * \param maxAllocation The desired allocation in us for each allocation period.
+   * \return The allocation duration for the allocation period.
+   */
+  virtual uint32_t GetAllocationDuration (uint32_t minAllocation, uint32_t maxAllocation);
```
description? it is trivial, but i would add it for the sake of completeness

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,108 @@
+   */
+  virtual uint32_t GetAllocationDuration (uint32_t minAllocation, uint32_t maxAllocation);
+  /**
+   * Implement the policy that accept, reject a new ADDTS request.
```
```suggestion
   * Implement the policy that accepts or rejects a new ADDTS request.
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,108 @@
+   */
+  virtual StatusCode AddNewAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info);
+  /**
+   * Implement the policy that accept, reject a modification request.
```
```suggestion
   * Implement the policy that accepts or rejects a modification request.
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,108 @@
+   * \param sourceAid The AID of the requesting STA.
+   * \param dmgTspec The DMG Tspec element of the ADDTS request.
+   * \param info The DMG Allocation Info element of the request.
+   * \return The Status Code to be included in the ADDTS response.
```
`StatusCode` or `status code`

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,108 @@
+
+private:
+  /**
+   * Verify how many blocks (in our case, one SP correspond to one block) we can guarantee to a periodic request.
```
```suggestion
   * Verify how many blocks (in our case, one SP corresponds to one block) we can guarantee to a periodic request.
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,108 @@
+   * Update the list of available time slots in the DTI.
+   * \param startAllocation start time of the allocation that has to be excluded from the available time.
+   * \param endAlloc end time of the allocation that has to be excluded from the available time.
+   * \param difference if not equal to 0, it represents how much an allocation has been reduced.
```
probably need to expand on this. you are assuming you can only reduce pre-existing allocations, right? this and other assumptions should probably be added directly into the header file as part of the documentation

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+        }
+      else
+        {
+          // if we cannot guarantee AT LEAST TWO periodic SPs, the request is rejected.
```
Should we check instead according to the Minimum Duration parameter in the DMG Tspec (I don't remember what we said but we discussed this). In any case, you can multiply the number of available blocks `blocks` by `allocDuration` and if this value satisfies Minimum Duration accept the request otherwise reject.

MatteoDrago replied:
This is also a good point and yes, we discussed about this offline. 
I will implement also this check, as you suggest :)

tommyazz replied:
In addition, should we check the number of blocks returned with the minimum allocation field in case of Isochronous traffic? If it's not possible to have more than one block again then reject

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+            {
+              // go on until eventually find the first available slot that fits this SP
+              // note that this also cover the condition where no slot satisfies the requirement
+              ++it;
```
Should you update also `startAlloc`? `continue` will start a new iteration of the while cycle (i.e. confronting `it` with `m_availableSlots.end ()`) but then `startAlloc` is for sure less than `it->first` and the cycle breaks.

MatteoDrago replied:
You are right. Here we should assign to `startAlloc` the new value of `it->first` to proceed with the check of the next slot in the list! How does this sound?

tommyazz replied:
Yeah, this was the fix I was hoping to hear ;)

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+        }
+    }
+
+  return blocks;
```
I think I got what this function is doing but it's quite complicated even if well commented. Maybe we should test this outside ns-3

MatteoDrago replied:
It is indeed complicated (even to design). I agree that we could test it outside ns-3. 
How do you suggest to do it?

tommyazz replied:
Let's discuss offline with @mattia-lecci

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+}
+
+void
+PeriodicDmgWifiScheduler::UpdateAvailableSlots (uint32_t startAllocation, uint32_t endAlloc, uint32_t difference)
```
Use the same name format for the first two parameters: startAlloc, endAlloc or startAllocation, endAllocation

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+{
+  NS_LOG_FUNCTION (this);
+
+  std::vector<std::pair<uint32_t, uint32_t> > newDTI;
```
Should be `newDti`

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+  else
+    {
+
+      if (m_availableSlots.begin () == m_availableSlots.end ())
```
Should this if-statement be called before `if (allocPeriod != 0)` ?. If we don't have available slots neither a periodic allocation can be allocated.

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+      if (m_availableSlots.begin () == m_availableSlots.end ())
+        {
+          NS_LOG_DEBUG ("There are no free available slots in the DTI.");
+          status.SetFailure ();
```
Should we call `return` immediately?

MatteoDrago replied:
yes, we should

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+              UpdateAvailableSlots (it->first, endAlloc);
+              status.SetSuccess ();
+              break;
+            }
```
if a requested allocation is Isochronous should we check also if the MinimumAllocation field is less than the slot duration before rejecting? As in the `BasicDmgWifiScheduler`

MatteoDrago replied:
we probably should, I'll update also this

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+  uint32_t startSlot = 0;
+  std::vector<std::pair<uint32_t, uint32_t> > newDTI; 
+
+  for (const auto & allocation: addtsListCopy)
```
What happens if one periodic allocation remains in the `m_addtsAllocationList`?
We would have to add many new available slots (according to the periodicity of this allocation), instead according to this code basically two available slots are added, one of them overlaps with the existing periodic allocation. Am I wrong?

MatteoDrago replied:
Oh God, this is a bad bug... It has to be fixed

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+        }
+    }
+
+  if (allocation == m_addtsAllocationList.end ())
```
Ideally  this check should not be necessary

MatteoDrago replied:
Why? What if the requested allocation does not exist?

tommyazz replied:
This should already be accounted for in `DmgWifiScheduler` if the allocation is not already present the method `ModifyExistingAllocation` is not called, `AddNewAllocation` is called instead. 
TBH, we can keep it as second layer of protection

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+
+  std::vector<std::pair<uint32_t, uint32_t> > newDTI;
+
+  if (difference > 0)
```
I would like to discuss the `if (difference > 0)` part of this function

MatteoDrago replied:
ok no prob

tommyazz reviewed (changes requested):
Great work Matteo! 
I'm asking some questions in order to better understand your code. I also think some parts need to be fixed but overall the code is good! Thanks for having written many comments, they were helpful.

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,108 @@
+   */
+  void UpdateAvailableSlots (uint32_t startAllocation, uint32_t endAlloc, uint32_t difference = 0);
+
+  std::vector<std::pair<uint32_t, uint32_t> > m_availableSlots;       //!< List of available time chunks in the DTI.
```
add what is the purpose of `std::pair<uint32_t, uint32_t>`, what it represents. I'm assuming start and end time of each available slot but it's just an assumption

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+    {
+      // no existing allocations
+      m_remainingDtiTime = m_dtiDuration.GetMicroSeconds ();
+      m_availableSlots.push_back (std::make_pair (0, m_dtiDuration.GetMicroSeconds ()));
```
so this should be the only element in `m_availableSlots` now, right? Are you sure this is what actually happens? if not, clear it first, if so, add an assert just to be sure

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+      m_remainingDtiTime = 0;
+      for (const auto & slot: m_availableSlots)
+        {
+          m_remainingDtiTime += (slot.second - slot.first);
```
`m_availableSlots`'s bounds are inclusive? if so, you should add 1. Not that it would change much but it would be the correct formula :)

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,524 @@
+PeriodicDmgWifiScheduler::GetAllocationDuration (uint32_t minAllocation, uint32_t maxAllocation)
+{
+  NS_LOG_FUNCTION (this << minAllocation << maxAllocation);
+  return ((minAllocation + maxAllocation) / 2);
```
i would return the max for the moment instead, i.e., what the STA really wants. We can tune this parameter in an attribute or a subclass later on

MatteoDrago committed:
Minor changes from TA's review

mattia-lecci reviewed (commented):
just a partial review, i'll finish it asap

MatteoDrago committed:
Minor changes from ML's review

MatteoDrago committed:
Fix bug in AdjustExistingAllocations 

The previous version of the method didn't consider the case where one or more periodic allocations were left in m_addtsAllocationList. Now this condition is checked and dealt with accordingly.

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -79,7 +81,7 @@ PeriodicDmgWifiScheduler::UpdateStartAndRemainingTime (void)
       for (const auto & slot: m_availableSlots)
         {
-          m_remainingDtiTime += (slot.second - slot.first);
+          m_remainingDtiTime += (slot.second - slot.first + 1);
```
sorry i thought i removed all the comments related to this. this is actually incorrect: considering a starting time and a duration with end time=start+duration you implicitly exclude the end time, so there's no need for this. sorry about that

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+StatusCode
+PeriodicDmgWifiScheduler::ModifyExistingAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info)
+{
+  NS_LOG_FUNCTION (this);
```
at least `sourceAid` can be printed

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+  for (allocation = m_addtsAllocationList.begin (); allocation != m_addtsAllocationList.end ();)
+    {
+      if ((allocation->GetAllocationID () == info.GetAllocationID ())
+          && (allocation->GetSourceAid () == sourceAid) && (allocation->GetDestinationAid () == info.GetDestinationAid ()))
```
separate it into two lines

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+  if (allocation == m_addtsAllocationList.end ())
+    {
+      NS_FATAL_ERROR ("Required allocation does not exist.");
+    }
```
```suggestion
NS_ABORT_MSG_IF (allocation == m_addtsAllocationList.end (), "Required allocation does not exist.");
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+  uint32_t timeDifference;
+  if (newDuration > currentDuration)
+    {
+      NS_LOG_DEBUG ("This feature is not supported.");
```
Which feature? you don't see the if statement when debugging :)

MatteoDrago replied:
no prob, I'll modify accordingly :)

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+        }
+      else
+        {
+          UpdateAvailableSlots (startAlloc, endAlloc, timeDifference);
```
are you separating this case just to build up the code for the [TODO](https://github.com/signetlabdei/ns3-802.11ad/blob/d92b1e8322237474816521f7e08ba79f2aeeb1ea/src/wifi/model/periodic-dmg-wifi-scheduler.cc#L485-L486)? if so, are you planning to solve it before the end of the PR?

MatteoDrago replied:
No, I think this is not fundamental in the current PR. The case is separated because I have to consider periodic and non-periodic allocations distinctly. 
I am not sure if I got your point

mattia-lecci replied:
similarly to what i wrote in another comment, non periodic allocations can be seen as an edge case of periodic ones, with a single block, so (i think) `allocation->GetNumberOfBlocks ()` at line 487 should return 1 and thus the for loop would work for that case as well.
With the TODO at line 485 you are saying that in future works you might want to add more blocks if possible, which would then be a significant change w.r.t. the non periodic allocations (even though it could still be easily managed).
So, did you separate the non periodic case because of that, just logically, or for some other reason? You know I'm not a fan of writing the same code twice, even if it's just a single line :)

MatteoDrago replied:
Now I understand what you meant. I decided to update as you suggested, leaving of course the TODO there to avoid forgetting about that as well.
Since the same could be said w.r.t. `AdjustExistingAllocations` I updated also that method accordingly. As you said, for non-periodic allocations the number of blocks is set to 1 and the allocation block period to 0 (it is easily verifiable in `DmgWifiScheduler`) so our solution covers all the possible conditions. 
Thanks for raising this point!

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+void
+PeriodicDmgWifiScheduler::UpdateAvailableSlots (uint32_t startAlloc, uint32_t endAlloc, uint32_t difference)
+{
+  NS_LOG_FUNCTION (this);
```
log input params

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+
+  std::vector<std::pair<uint32_t, uint32_t> > newDti;
+
+  if (difference > 0)
```
the fact that you use the same function for two different operations is a bit confusing to me. Either document it better or separate into 2 separate functions (e.g., overloading it with 2 and 3 input arguments). The body of the two subfunctions is completely independent..
Also, i would actually propose to use `int64_t` to support both positive and negative difference, even though for the moment it is not really necessary.

MatteoDrago replied:
I'll go for the overloading proposal, good point. 
I don't think that it is necessary for now to use `int64_t`. It could actually be misleading as we would need to check the sign and return an error in case a negative value has been passed (as now the increase in duration is not supported).

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+
+          if (endAlloc < slot.first)
+            {
+              // the following snippet ensure that the search is carried on until 
```
```suggestion
              // the following snippet ensures that the search is carried on until 
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+    {
+      // something has changed in the allocation list, need to change the list of 
+      // available slots accordingly
+      bool search = true;
```
i would call it something like `keepSearching` but it's just a preference. i have to read some code to understand what `search` does

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+                      newDti.push_back (std::make_pair (endAlloc, endAlloc + difference));
+                      newDti.push_back (slot);
+                      search = false;
+                    }
```
i don't quite get this part, could you explain, maybe offline?

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+              newDti.push_back (std::make_pair (slot.first, startAlloc));
+              newDti.push_back (std::make_pair (endAlloc, slot.second));
+            }
+        }
```
i think start/end times are inconsistent: is the allocation of the type
1. [start,end]
2. [start,end)
3. (start,end]
4. (start,end)
I think think types 1 and 2 are the most common and more intuitive but you need to be careful.
* Type 1 requires endAlloc=startAlloc+duration-1, but then you can say that two contiguous allocations will be
[start1,end1], [start2,end2]
with start2==end1+1, so no conflicts on a shared us (see type 2 for a clarification).
* Type 2, instead, requires endAlloc=startAlloc+duration, but then you can say that two contiguous allocations will be
[start1,end1], [start2,end2]
with start2==end1.
This might create problems when sending the allocation list to the STAs since for a us both allocations will be active, but I guess this is one of the problems that the guard time solves (?)

Looking at this code you are assuming type 2. Warning: are guard times included somewhere else?

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+        }
+    }
+  else
+    {
```
add a short comment on that difference==0 means, it took me a while to understand that this function does 2 jobs at once :)

mattia-lecci replied:
actually, since the two cases are well separated, why not separating them in two different functions? Maybe overloading this one in a 2 and 3 input arguments versions

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -0,0 +1,113 @@
+
+  // std::pair<uint32_t, uint32_t> is a struct used to store the start and 
+  // end time (first and second member of the pair, respectively) of the available time chunks
+  std::vector<std::pair<uint32_t, uint32_t> > m_availableSlots;       //!< List of available time chunks in the DTI.
```
maybe `std::vector` is not the best choice in this case: most of the times you are modifying elements in the middle of the vector without knowing their index, which requires O(n) operations to find/insert/remove. Probably `std::list` is more appropriate in this case.
If you are maintaining the order, you might want to think about using `std::set` instead: find/insert/remove are all O(log(n)) operations. If you design it with some care you can even find the desired slot in a single line (`find`, `lower_bound`, `upper_bound`), thus avoid complicated and long loops, but this is a bit more tricky

MatteoDrago replied:
From this point of view, my solution is the simplest possible. No particular optimization techniques have been implemented :) You raised a good point, I'll try and see what can be done to optimize as you suggested

mattia-lecci replied:
if you want to get over with this quickly (which i would totally agreed on), just using `std::list` instead of `std::vector` would suffice. It's the brute force way, but at least you use the proper data structure for that :)

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+{
+  NS_LOG_FUNCTION (this);
+
+  std::vector<std::pair<uint32_t, uint32_t> > newDti;
```
can't you operate in-place on `m_availableSlots` directly? you would avoid recreating a new vector of pairs and only modify/insert/remove the required slots

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+{
+  NS_LOG_FUNCTION (this);
+
+  auto it = m_availableSlots.begin ();
```
any reason why you chose while instead of for loop? not that it really matters, but it would be one less line and a scoped iterator :)

MatteoDrago replied:
it is easier to remain in the same slot when it is possible to add more than one SP in the same slot! I found it more easier to implement

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+
+  auto it = m_availableSlots.begin ();
+  uint32_t startAlloc = it->first;
+  uint32_t slotDuration = 0;
```
`remainingSlotDuration`?

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+        {
+          // pass to the next slot only if startAlloc goes beyond the end of the current one
+          ++it;
+        }
```
There's no need to check these things, you will do it during the next loop. You are basically copying the code from the previous if statement in a slightly different way. If later on you find a bug you'll have to fix both conditional statements: chances are that you (or anybody) will forget about it and replace a bug with another one, probably more tricky to find

```suggestion
      // if next allocation period exceeds the current available slot's boundaries
      // proceed to the next one
      if (startAlloc > it->second)
        {
          ++it;
        }
      // else: the next allocation period is still (partially) in the current
      // available slot
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+        {
+          // Periodicity has been broken, we stop the algorithm.
+          break;
+        }
```
i understood this condition after a while, can we improve the comment so that its purpose is immediately (or almost) clear?

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+
+      blocks++;
+      startAlloc += spInterval;
+
```
Only an octet is reserved for Number of Blocks, avoid overflows when BI duration is extremely long or the periodicity is extremely short

```suggestion
      if (blocks == 255)
        {
          // Number of blocks described by an octet: only up to 255 blocks
          break;
        }
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+StatusCode
+PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info)
+{
+  NS_LOG_FUNCTION (this);
```
log at least sourceAid

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+  StatusCode status;
+  uint32_t allocDuration;
+
+  if (m_availableSlots.begin () == m_availableSlots.end ())
```
```suggestion
  if (m_availableSlots.empty ())
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
+    {
+      allocDuration = GetAllocationDuration (dmgTspec.GetMinimumAllocation (), dmgTspec.GetMaximumAllocation ());
+      if (allocDuration < dmgTspec.GetMinimumAllocation ())
+        {
+          NS_LOG_DEBUG ("Unable to guarantee minimum duration.");
+          status.SetFailure ();
+          return status;
+        }
```
i guess you're thinking of children classes extending this one with custom `GetAllocationDuration`?

MatteoDrago replied:
Of course

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+      return status;
+    }
+
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
```
why isn't this a scoped enum? :'(
i'm not complaining about your code

MatteoDrago replied:
I don't get what you mean, anyway! How could this be improved by scoped enum?

mattia-lecci replied:
https://stackoverflow.com/questions/18335861/why-is-enum-class-preferred-over-plain-enum

You would use `AllocationFormat::ISOCHRONOUS` instead of `ISOCHRONOUS`. It's safer to use and avoids some common bugs/programming mistakes, especially for projects this large

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+
+      // spInterval is going to be passed to AddAllocationPeriod to specify the 
+      // distance between consecutive periodic SPs
+      uint32_t spInterval = uint32_t (m_biDuration.GetMicroSeconds () / allocPeriod);
```
https://www.nsnam.org/develop/contributing-code/coding-style/#Casts
here brace initialization cannot be used since you're losing information (uint64_t -> uint32_t), so you should probably use a static cast

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+      // distance between consecutive periodic SPs
+      uint32_t spInterval = uint32_t (m_biDuration.GetMicroSeconds () / allocPeriod);
+
+      NS_LOG_DEBUG ("Allocation Period " << uint32_t (allocPeriod) << " AllocDuration " << allocDuration << " Multiple " << isMultiple);
```
you should be able to print `uint16_t` directly
```suggestion
      NS_LOG_DEBUG ("Allocation Period " << allocPeriod << " AllocDuration " << allocDuration << " Multiple " << isMultiple);
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+      uint32_t spInterval = uint32_t (m_biDuration.GetMicroSeconds () / allocPeriod);
+
+      NS_LOG_DEBUG ("Allocation Period " << uint32_t (allocPeriod) << " AllocDuration " << allocDuration << " Multiple " << isMultiple);
+      NS_LOG_DEBUG ("Schedule one SP every " << spInterval);
```
i really don't like to have multiple debug lines from the same function, so i would avoid them whenever possible

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+                               startFirstAllocation, allocDuration, spInterval, blocks);
+          status.SetSuccess ();
+        }
+      else
```
this is an edge case of a periodic allocation, can't we join them together? again, for code reuse and single point of failure.
E.g., `GetAvailableBlocks` could be in common, for example adding a `maxBlocks` input which defaults to `255` (8 bits, see related comment) and could be, instead, `1` for a non-periodic allocation.

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+              counter++;
+
+              // update remaining DTI time for consistency
+              m_remainingDtiTime -= (allocDuration + m_guardTime);
```
can't this be part of `UpdateAvailableSlots`? there's a strong risk of forgetting to update it otherwise

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+
+              endAlloc = AllocateSingleContiguousBlock (info.GetAllocationID (), info.GetAllocationType (), info.IsPseudoStatic (),
+                                                        sourceAid, info.GetDestinationAid (), it->first, allocDuration);
+              m_remainingDtiTime -= (allocDuration + m_guardTime);
```
same as above, i almost thought you forgot about it but it was just in a different position :)

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,539 @@
+          uint8_t blocks = allocation.GetNumberOfBlocks ();
+          while (blocks > 0)
+            {
+              UpdateAvailableSlots (startAlloc, endAlloc);
```
don't you need to update `m_remainingDtiTime`?

mattia-lecci replied:
there are more points in the code where i have this doubt related to `UpdateAvailableSlots`, please look at the previous comment

mattia-lecci reviewed (commented):
now i get the amount of work you had to do.. great job! :)

MatteoDrago committed:
First round of minor changes from Mattia's review

MatteoDrago committed:
Overload of UpdateAvailableSlots

Now the method is divided into two distinct versions that address two different conditions, i.e., new allocation added to the list of allocations and a change in previously allocated slots.

MatteoDrago committed:
Update consistently m_remainingDtiTime and avoid redundant code

MatteoDrago committed:
Fix GetAvailableBlocks and improve doxy

MatteoDrago committed:
Modify GetAvailableBlocks and avoid redundant code in AddNewAllocation

GetAvailableBlocks has been modified to return a list containing the starting time of each block that can be allocated. Accordingly, AddNewAllocation has been modified to consider a non-periodic allocation has an edge case of the periodic one.

MatteoDrago committed:
Optimize UpdateAvailableSlot

m_availableSlots has been changed from vector to list and both versions of UpdateAvailableSlots have been updated to perform in-place insertion/deletion.

MatteoDrago commented:
I also attach the final output from the latest version of the periodic scheduler:
![Screenshot from 2020-05-18 16-13-22](https://user-images.githubusercontent.com/18742612/82223287-c1591500-9922-11ea-8187-4480937dcf28.png)

obtained running `evaluate_qd_channel_lroom_scenario.cc` with the following allocations:
![Screenshot from 2020-05-18 16-13-52](https://user-images.githubusercontent.com/18742612/82223295-c322d880-9922-11ea-9815-ae4c920bf6ee.png)

mattia-lecci commented:
> I also attach the final output from the latest version of the periodic scheduler:
> ![Screenshot from 2020-05-18 16-13-22](https://user-images.githubusercontent.com/18742612/82223287-c1591500-9922-11ea-8187-4480937dcf28.png)

Is there a bug with the APP layer TX/RX packets?

MatteoDrago commented:
> > I also attach the final output from the latest version of the periodic scheduler:
> > ![Screenshot from 2020-05-18 16-13-22](https://user-images.githubusercontent.com/18742612/82223287-c1591500-9922-11ea-8187-4480937dcf28.png)
> 
> Is there a bug with the APP layer TX/RX packets?

If I recall correctly, yes, there is a bug. We already discussed this in a gdoc, if I'm not wrong

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.h:
```diff
@@ -88,23 +88,32 @@ class PeriodicDmgWifiScheduler : public DmgWifiScheduler
-   * \param spInterval time between two consecutive periodic SPs.
-   * \return number of blocks that can be allocated in the DTI
+   * \param spInterval time between two consecutive periodic SPs (= 0 if single non-periodic request).
+   * \param maxBlocksNumber maximum numbero of blocks allowed (= MAX_NUM_BLOCKS if periodic, = 1 if non-periodic)
```
lol
```suggestion
   * \param maxBlocksNumber maximum number of blocks allowed (= MAX_NUM_BLOCKS if periodic, = 1 if non-periodic)
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -178,121 +173,95 @@ PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecEle
-      uint32_t endAlloc;
-      uint8_t counter = 0;
+      NS_LOG_DEBUG ("Allocation Period " << allocPeriod 
+      << " AllocDuration " << allocDuration 
+      << " - Schedule one SP every " << spInterval);
```
align

MatteoDrago replied:
it is more readable from the terminal in this format

mattia-lecci replied:
does check-style.py agree?

MatteoDrago replied:
now I get what you mean, I misunderstood sorry

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -178,121 +173,95 @@ PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecEle
+      UpdateAvailableSlots (startAlloc, endAlloc);
 
+      startAlloc += spInterval;
+      counter++;
```
reuse what you already computed: single point of failure
```suggestion
  uint32_t endAlloc;
  for (i = 0; i < blocks.size(); ++i)
    {
      NS_LOG_DEBUG ("Reserve from " << blocks[i] << " for " << allocDuration);

      endAlloc = blocks[i] + allocDuration + m_guardTime;
      UpdateAvailableSlots (blocks[i], endAlloc);
    }
```

MatteoDrago replied:
true, I didn't think to this possible improvement

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -178,121 +173,95 @@ PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecEle
+PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t spInterval, uint8_t maxBlocksNumber)
 {
-  NS_LOG_FUNCTION (this);
+  NS_LOG_FUNCTION (this << allocDuration << spInterval);
```
```suggestion
  NS_LOG_FUNCTION (this << allocDuration << spInterval << +maxBlocksNumber);
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -178,121 +173,95 @@ PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecEle
-          // Periodicity has been broken, we stop the algorithm.
+          // The next periodic SP should be positioned before the beginning of the 
+          // next available slot: this translates in a broken periodicity, and the 
+          // algorithm stops.
```
```suggestion
          // The next periodic SP block is positioned before the beginning of the 
          // next available slot: this translates in a broken periodicity, and the 
          // algorithm stops.
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -335,79 +293,97 @@ PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t s
-      // available slots accordingly
-      bool search = true;
-      for (const auto & slot: m_availableSlots)
+      startSlot = (*it).first;
+      endSlot = (*it).second;
```
```suggestion
      startSlot = it->first;
      endSlot = it->second;
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -335,79 +293,97 @@ PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t s
-            }
+      if (startSlot == startAlloc)
+        {
+          m_availableSlots.insert(it, std::make_pair (endAlloc, endSlot));
+          it = m_availableSlots.erase(it);
```
Doesn't this work? It would be easier to read and more efficient as well (need reference iterator)
```suggestion
          it->first = endAlloc;
          ++it;
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -335,79 +293,97 @@ PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t s
+  uint32_t endSlot;
 
-  if (difference > 0)
+  for (auto it = m_availableSlots.begin(); it != m_availableSlots.end(); ++it)
```
advancing the iterator in the for loop while inserting/erasing can lead to unexpected results. it's safer do to it manually (yes, you could use a while loop :) )
```suggestion
  for (auto& it = m_availableSlots.begin(); it != m_availableSlots.end();)
```

MatteoDrago replied:
I specifically did it in this way because I know what I am modifying and how the iterator would change according to any insertion/deletion. Also, remember that once the algorithm finds the slot that needs to be changed, it proceeds with the change and then stops. 
I would stick with my solution, with some of the improvements that you suggested (such as the referenced iterator).

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -335,79 +293,97 @@ PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t s
+      
+      if (startSlot > endAlloc || endSlot < startAlloc)
         {
+          continue;
```
```suggestion
          ++it;
          continue;
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -335,79 +293,97 @@ PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t s
+        }
+      else if (startSlot < startAlloc && endSlot > endAlloc)
+        {
+          m_availableSlots.insert(it, std::make_pair (startSlot, startAlloc));
+          m_availableSlots.insert(it, std::make_pair (endAlloc, endSlot));
+          it = m_availableSlots.erase(it);
```
```suggestion
          it->first = endAlloc;
          it = m_availableSlots.insert(it, std::make_pair (startSlot, startAlloc)); // insert before
          ++it; // going to the [endAlloc, endSlot] slot
          ++it; // next unseen slot
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -335,79 +293,97 @@ PeriodicDmgWifiScheduler::GetAvailableBlocks (uint32_t allocDuration, uint32_t s
+}
+
+void
+PeriodicDmgWifiScheduler::UpdateAvailableSlots (uint32_t startAlloc, uint32_t newEndAlloc, uint32_t difference)
```
same suggestions as for the 2-input one

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -476,30 +450,21 @@ PeriodicDmgWifiScheduler::ModifyExistingAllocation (uint8_t sourceAid, const Dmg
+          startAlloc += allocPeriod;
+          endAlloc += allocPeriod;
+          blocks -= 1;
         }
```
why do you like while loops so much? :P

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -111,27 +111,22 @@ PeriodicDmgWifiScheduler::AdjustExistingAllocations (AllocationFieldListI iter,
-      else
+      // if the number of blocks allocated is > 1, the allocation is periodic 
+      uint8_t blocks = allocation.GetNumberOfBlocks ();
+      while (blocks > 0)
```
damn while loop :P

mattia-lecci reviewed (changes requested):
be careful when modifying a list inside a for loop, it can lead to nasty bugs.
Also, please override all virtual methods

MatteoDrago committed:
Optimize code structure and fix typos, ML's review

MatteoDrago committed:
Forgot to save previous updates

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+      m_remainingDtiTime = m_dtiDuration.GetMicroSeconds ();
+      // clear the list of available slots: if no allocations have been scheduled, then the DTI is completely available
+      m_availableSlots.clear ();
+      m_availableSlots.push_back (std::make_pair (0, m_dtiDuration.GetMicroSeconds ()));
```
```suggestion
      m_availableSlots.push_back (std::make_pair (0, m_remainingDtiTime));
```

MatteoDrago replied:
good catch

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+  m_availableSlots.clear ();
+  m_availableSlots.push_back (std::make_pair (0, m_dtiDuration.GetMicroSeconds ()));
+  // reset m_remainingDtiTime as well
+  m_remainingDtiTime = m_dtiDuration.GetMicroSeconds ();
```
Maybe you can set `m_remainingDtiTme` before calling `m_availableSlots.push_back` and use it. Like in `UpdateStartAndRemainingTime`

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+{
+  NS_LOG_FUNCTION (this << startAlloc << endAlloc);
+
+  uint32_t startSlot;
```
In one line, like you did in other parts of this code ;)

MatteoDrago replied:
I didn't get what you mean here

tommyazz replied:
```suggestion
  uint32_t startSlot, endSlot;
```

MatteoDrago replied:
sorry for my dumbness, I didn't get it before

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
+    {
+      allocDuration = GetAllocationDuration (dmgTspec.GetMinimumAllocation (), dmgTspec.GetMaximumAllocation ());
+      if (allocDuration < dmgTspec.GetMinimumAllocation ())
```
I don't think this check should be here. Ideally `GetAllocationDuration` returns a value between Minimum Allocation and Maximum Allocation. For the time being is just the mean between those two value. See comment below.

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+
+      blocks = GetAvailableBlocks (allocDuration, spInterval, MAX_NUM_BLOCKS);
+
+      if (blocks.size () <= 1)
```
If we are not able to get a number of blocks > 1 by giving to `GetAvailabeBlocks` the allocation duration computed in `GetAllocationDuration`, it is worth trying to give to this method the minimum allocation duration, provided that this value is different from the one returned by `GetAllocationDuration` (the former should be less). 
Maybe the minimum allocation duration fits a greater number of blocks and we are able to accept the allocation.
This holds just in the case of Isochronous traffic.

MatteoDrago replied:
here you can literally do whatever you want with the fine-tuning of the AllocationDuration. I just did the simplest possible, as discussed :) If you think it is worth adding this check and give it a try, we can do it.

mattia-lecci replied:
i agree on keeping it simple. this (and many other heuristics) are part of our research on SP scheduling and thus of our future works

tommyazz replied:
Ok I see. The thing is that this feature is implemented in the `BasicDmgWifiScheduler`. I thought we should make this `PeriodicDmgWifiScheduler` equal the Basic when dealing with non-periodic traffic. 
In any case, I have proposed some code if we want to implement this (it's just a few lines of more code)

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+  else
+    {
+      blocks = GetAvailableBlocks (allocDuration, 0, 1);
+      if (blocks.size () == 0)
```
Same comment as before.

tommyazz replied:
Proposed code for this part is almost equal to the proposed code above

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+  m_allocationList = m_addtsAllocationList;
+  AllocationFieldList broadcastCbapList;
+
+  // fill all the remaining available slots with broadcast CBAPs
```
At this point, are `m_availableSlots` and `m_addtsAllocationList`already ordered? 
Is there a case when they can be unordered? If no, I think we can avoid the sort at the end and insert elements in `m_allocationsList` using an incrementing iterator.

MatteoDrago replied:
`m_addtsAllocationList` can be unordered, as discussed offline, and this is reflected to `m_allocationList`. By construction, instead, `m_availableSlots` is always ordered, which is the reason why we can do the for-loop in this method. 
The sort at the end of the function is done just for my own mental sanity: when debugging, it is easier to read an ordered list in order to understand if everything is working properly. Having said that, yes, if you want we can delete it.

tommyazz replied:
Sure, you told me but I forgot the reason why, now I remember thanks :)

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+
+  uint32_t currentDuration = allocation->GetAllocationBlockDuration ();
+  NS_LOG_DEBUG ("current duration=" << currentDuration << ", new duration=" << newDuration);
+  uint32_t timeDifference;
```
In the case of Isochronous traffic, we should check also the minimum allocation duration before aborting the modification request. Maybe the minimum allocation value is less then the current duration.

tommyazz reviewed (changes requested):
Hi Matteo,
I really like the actual code, but I think there's one more thing to consider in the case of isochronous requests. Please have a look at my comments

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+  NS_LOG_FUNCTION (this << +sourceAid);
+
+  StatusCode status;
+  uint32_t allocDuration;
```
```suggestion
  uint32_t allocDuration, minimumAllocation;
```

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
+    {
+      allocDuration = GetAllocationDuration (dmgTspec.GetMinimumAllocation (), dmgTspec.GetMaximumAllocation ());
```
```suggestion
      allocDuration = GetAllocationDuration (dmgTspec.GetMinimumAllocation (), dmgTspec.GetMaximumAllocation ());
      minimumAllocation = dmgTspec.GetMinimumAllocation ();
```

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+  else if (info.GetAllocationFormat () == ASYNCHRONOUS)
+    {
+      // for asynchronous allocations, the Maximum Allocation field is reserved (IEEE 802.11ad 8.4.2.136)
+      allocDuration = dmgTspec.GetMinimumAllocation ();
```
```suggestion
      allocDuration = dmgTspec.GetMinimumAllocation ();
      minimumAllocation = allocDuration;
```

tommyazz commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -0,0 +1,490 @@
+
+      if (blocks.size () <= 1)
+        {
+          // if we cannot guarantee AT LEAST TWO periodic SPs, the request is rejected
```
```suggestion
          if (info.GetAllocationFormat () == ISOCHRONOUS && minimumAllocation < allocDuration)
            {
              /* Get number of available blocks using minimum allocation duration */
              blocks = (minimumAllocation, spInterval, MAX_NUM_BLOCKS); 
              if (blocks.size () <= 1)
               {
                  // if we cannot guarantee AT LEAST TWO periodic SPs, the request is rejected
                  status.SetFailure ();
                  return status;
               }
            }
          else
            {
              status.SetFailure ();
              return status;
            }
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -120,13 +120,12 @@ PeriodicDmgWifiScheduler::AdjustExistingAllocations (AllocationFieldListI iter,
       // if the number of blocks allocated is > 1, the allocation is periodic 
       uint8_t blocks = allocation.GetNumberOfBlocks ();
-      while (blocks > 0)
+      for (uint8_t i = 0; i < blocks; ++i)
```
```suggestion 
      for (uint8_t i = 0; i < allocation.GetNumberOfBlocks (); ++i)
```

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -300,31 +292,29 @@ PeriodicDmgWifiScheduler::UpdateAvailableSlots (uint32_t startAlloc, uint32_t en
   uint32_t endSlot;
 
-  for (auto it = m_availableSlots.begin(); it != m_availableSlots.end(); ++it)
+  for (auto it = m_availableSlots.begin (); it != m_availableSlots.end (); ++it)
```
does this code work even without passing the iterator by ref?

MatteoDrago replied:
it didn't even compile with `auto&` as you suggested. This version compiles and works (usual test, checking the allocations)

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -300,31 +292,29 @@ PeriodicDmgWifiScheduler::UpdateAvailableSlots (uint32_t startAlloc, uint32_t en
+          it->first = endAlloc;
+          m_availableSlots.insert (it, std::make_pair (startSlot, startAlloc));
           break;
         }
```
what if `endSlot == endAlloc`? It would probably enter the last condition and create a zero-duration slot, right?

mattia-lecci replied:
also, just to be sure i would add an `else` with fatal error. this holds for all if statements, especially when they are long and complicated, so it's possible that (i) you miss one edge case, (ii) some runtime error breaks one of your assumptions

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -451,18 +438,17 @@ PeriodicDmgWifiScheduler::ModifyExistingAllocation (uint8_t sourceAid, const Dmg
       // add further blocks
       uint8_t blocks = allocation->GetNumberOfBlocks ();
-      while (blocks > 0)
+      for (uint8_t i = 0; i < blocks; ++i)
```
```suggestion
      for (uint8_t i = 0; i < allocation->GetNumberOfBlocks (); ++i)
```

mattia-lecci reviewed (commented):
just a couple of minor comments :)

MatteoDrago committed:
Addressed review comments

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -191,24 +192,49 @@ PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecEle
+            }
+          else
+            {
+              status.SetFailure ();
```
so if it's not isochronous it fails? is it what you meant or is it a bug?

mattia-lecci commented on src/wifi/model/periodic-dmg-wifi-scheduler.cc:
```diff
@@ -191,24 +192,49 @@ PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecEle
+            }
+          else
+            {
+              status.SetFailure ();
```
same as above

mattia-lecci reviewed (commented):
last check before merge :)
